src/core/eggmath.rkt

```diff
@@ -0,0 +1,213 @@
+                    (herbie-pattern->rust-pattern (rule-output rule)))))
+
+(define (egraph-run-iter egraph-data node-limit ffi-rules precompute?)
+  (egraph_run_iter (egraph-data-egraph-pointer egraph-data) node-limit ffi-rules precompute?))
+
+
+(define (egraph-run-rules egraph-data node-limit rules precompute?)
+  (define ffi-rules (make-ffi-rules rules))
+  (define old-cnt 0)
+  (for/and ([iter (in-naturals 0)])
+    (egraph_run_iter (egraph-data-egraph-pointer egraph-data) node-limit ffi-rules precompute?)
+    (define cnt (egraph_get_size (egraph-data-egraph-pointer egraph-data)))
+    (define is_stop (or (>= cnt node-limit) (<= cnt old-cnt)))
+    (set! old-cnt cnt)
+    (not is_stop)))
+
+
+;; calls the function on a new egraph, and cleans up
+(define (egraph-run egraph-function)
+  (define egraph (egraph-data (egraph_create) (make-hash) (make-hash)))
+  (define res (egraph-function egraph))
+  (egraph_destroy (egraph-data-egraph-pointer egraph))
+  res)
+
+(define (egg-expr->expr expr eg-data)
+  (define parsed (read (open-input-string expr)))
+  (egg-parsed->expr parsed (egraph-data-egg->herbie-dict eg-data)))
+
+(define (egg-parsed->expr parsed rename-dict)
+  (cond
+    [(list? parsed)
+     (cons
+      (first parsed)
+      (for/list ([expr (rest parsed)])
+        (egg-parsed->expr expr rename-dict)))]
+    [(and (number? parsed) (exact? parsed) (real? parsed))
+     parsed]
+    [else
+     (if (set-member? fpconstants parsed)
+         parsed
+         (hash-ref rename-dict parsed))]))
+
+;; returns a pair of the string representing an egg expr, and updates the hash tables in the egraph
+;; the hash table maps all symbols and non-integer values to new names for egg
+(define (expr->egg-expr expr egg-data)
+  (define egg->herbie-dict (egraph-data-egg->herbie-dict egg-data))
+  (define herbie->egg-dict (egraph-data-herbie->egg-dict egg-data))
+  (expr->egg-expr-helper expr egg->herbie-dict herbie->egg-dict))
+
+(define (expr->egg-expr-helper expr egg->herbie-dict herbie->egg-dict)
+  (cond
+    [(list? expr)
+     (string-append
+      "("
+      (symbol->string (first expr))
+      (foldr
+       (lambda (sub-expr acc)
+         (string-append " " (expr->egg-expr-helper sub-expr egg->herbie-dict herbie->egg-dict)
+                        acc))
+       "" (rest expr))
+      ")")]
+    [(and (number? expr) (exact? expr) (real? expr))
+     (number->string expr)]
+    [(hash-has-key? herbie->egg-dict expr)
+     (symbol->string (hash-ref herbie->egg-dict expr))]
+    [else
+     (define new-key (string-append "h" (number->string (hash-count herbie->egg-dict))))
+     (define new-key-symbol (string->symbol new-key))
+         
+     (hash-set! herbie->egg-dict
+                expr
+                new-key-symbol)
+     (hash-set! egg->herbie-dict
+                new-key-symbol
+                expr)
+      new-key]))
+
+
+(define-struct (egg-add-exn exn:fail:user) ())
```

Comment:
I don't think this is a user failure! `exn:fail` is a fine parent here, but `exn:fail:user` implies that the non-technical end-user did something wrong and hides the traceback.

Reply:
Also isn't `define-struct` deprecated?

Reply:
From [the docs](https://docs.racket-lang.org/reference/define-struct.html#%28form._%28%28lib._racket%2Fprivate%2Fbase..rkt%29._define-struct%29%29):

> This form is provided for backwards compatibility; struct is preferred.